File: Context/Sources/CodeFire/Services/ProjectDiscovery.swift
import Foundation
import GRDB

/// Discovers Claude Code projects from `~/.claude/projects/` and imports them
/// (along with their sessions) into the local database.
///
/// Claude Code stores per-project data in directories whose names encode the
/// absolute project path. Slashes, spaces, dots, and existing hyphens all
/// become `-`, making the encoding ambiguous. `ProjectDiscovery` resolves
/// each encoded name back to a real filesystem path using backtracking against
/// the actual directory structure on disk.
class ProjectDiscovery {

    // MARK: - Properties

    private let claudeDir: URL       // ~/.claude
    private let db: DatabaseService
    private let fileManager = FileManager.default

    // MARK: - Init

    init(db: DatabaseService = .shared) {
        self.db = db
        self.claudeDir = fileManager.homeDirectoryForCurrentUser
            .appendingPathComponent(".claude", isDirectory: true)
    }

    // MARK: - Public API

    /// Scan `~/.claude/projects/` and return a `Project` record for each
    /// directory whose encoded name can be resolved to a real path.
    func discoverProjects() throws -> [Project] {
        let projectsDir = claudeDir.appendingPathComponent("projects", isDirectory: true)

        guard fileManager.fileExists(atPath: projectsDir.path) else {
            return []
        }

        let entries = try fileManager.contentsOfDirectory(
            at: projectsDir,
            includingPropertiesForKeys: [.isDirectoryKey],
            options: [.skipsHiddenFiles]
        )

        var projects: [Project] = []

        for entry in entries {
            // Only look at directories
            guard (try? entry.resourceValues(forKeys: [.isDirectoryKey]))?.isDirectory == true
            else { continue }

            let encodedName = entry.lastPathComponent

            // Skip the bare `-` directory (represents `/`, not a real project)
            guard encodedName != "-" else { continue }

            // Try to resolve the encoded name back to a filesystem path
            guard let resolvedPath = resolvePath(from: encodedName) else {
                // Could not resolve — project directory may have been deleted.
                // Still create a record so the user can see historical sessions.
                let fallbackName = encodedName.split(separator: "-").last.map(String.init) ?? encodedName
                let project = Project(
                    id: UUID().uuidString,
                    name: fallbackName,
                    path: encodedName, // store encoded name as-is
                    claudeProject: entry.path,
                    lastOpened: nil,
                    createdAt: Date()
                )
                projects.append(project)
                continue
            }

            let name = (resolvedPath as NSString).lastPathComponent
            let project = Project(
                id: UUID().uuidString,
                name: name,
                path: resolvedPath,
                claudeProject: entry.path,
                lastOpened: nil,
                createdAt: Date()
            )
            projects.append(project)
        }

        return projects
    }

    /// Discover projects and insert any that are not already in the database
    /// (matched by `path`).
    func importProjects() throws {
        let discovered = try discoverProjects()

        try db.dbQueue.write { dbConn in
            for project in discovered {
                // Skip if a project with the same path already exists
                let exists = try Project
                    .filter(Project.Columns.path == project.path)
                    .fetchCount(dbConn) > 0

                if !exists {
                    var p = project
                    try p.insert(dbConn)
                }
            }
        }
    }

    /// Discover `.jsonl` session files for a project and import any that are
    /// not already in the database.
    func importSessions(for project: Project) throws {
        guard let claudeProjectPath = project.claudeProject else { return }

        let claudeProjectURL = URL(fileURLWithPath: claudeProjectPath, isDirectory: true)

        guard fileManager.fileExists(atPath: claudeProjectURL.path) else { return }

        let contents = try fileManager.contentsOfDirectory(
            at: claudeProjectURL,
            includingPropertiesForKeys: nil,
            options: [.skipsHiddenFiles]
        )

        let jsonlFiles = contents.filter { $0.pathExtension == "jsonl" }

        try db.dbQueue.write { dbConn in
            for file in jsonlFiles {
                let sessionId = file.deletingPathExtension().lastPathComponent

                // Check if session already exists
                let existing = try Session
                    .filter(Session.Columns.id == sessionId)
                    .fetchOne(dbConn)

                // Skip if session exists AND already has token data
                if let existing = existing, existing.inputTokens > 0 || existing.outputTokens > 0 {
                    continue
                }

                // Parse the session file (new session or backfill tokens)
                guard let parsed = try? SessionParser.parse(fileURL: file) else { continue }

                let summary = SessionParser.generateSummary(from: parsed)

                let filesChangedJSON: String? = {
                    guard !parsed.filesChanged.isEmpty,
                          let data = try? JSONEncoder().encode(parsed.filesChanged),
                          let str = String(data: data, encoding: .utf8)
                    else { return nil }
                    return str
                }()

                var session = Session(
                    id: parsed.sessionId,
                    projectId: project.id,
                    slug: parsed.slug,
                    startedAt: parsed.startedAt,
                    endedAt: parsed.endedAt,
                    model: parsed.model,
                    gitBranch: parsed.gitBranch,
                    summary: summary,
                    messageCount: parsed.messageCount,
                    toolUseCount: parsed.toolUseCount,
                    filesChanged: filesChangedJSON,
                    inputTokens: parsed.inputTokens,
                    outputTokens: parsed.outputTokens,
                    cacheCreationTokens: parsed.cacheCreationTokens,
                    cacheReadTokens: parsed.cacheReadTokens
                )
                try session.save(dbConn) // save = insert or update
            }
        }
    }

    // MARK: - Path Resolution

    /// The characters that Claude Code's encoding maps to `-`.
    /// A `-` in the encoded name could have originally been any of these.
    private static let replacementChars: [Character] = ["-", " ", "."]

    /// Maximum time (in seconds) to spend resolving a single encoded name
    /// before giving up. Prevents runaway backtracking for deleted projects
    /// with deeply ambiguous encodings.
    private static let resolveTimeoutSeconds: TimeInterval = 0.5

    /// Resolve an encoded Claude project directory name back to a real
    /// filesystem path.
    ///
    /// Claude Code encodes absolute paths by replacing `/`, ` `, `.`, (and
    /// keeping `-` as `-`) all with `-`. This function resolves the ambiguity
    /// by trying each possible original character at every `-` position and
    /// checking the filesystem.
    ///
    /// The algorithm works at two levels:
    /// 1. **Path component boundary**: at each `-`, try treating it as a `/`
    ///    and check if the prefix is a real directory.
    /// 2. **Within-component characters**: if `/` doesn't work, try `-`, ` `,
    ///    and `.` to continue building the current component name.
    ///
    /// Uses memoization on `(charIndex, parentDir)` to prune redundant
    /// branches, and a timeout to bound worst-case runtime.
    ///
    /// - Parameter encodedName: e.g. `-Users-nicknorris-Documents-my-project`
    /// - Returns: The resolved absolute path, or `nil` if no valid path is found.
    func resolvePath(from encodedName: String) -> String? {
        // Strip leading dash (represents the leading `/`)
        var stripped = encodedName
        if stripped.hasPrefix("-") {
            stripped = String(stripped.dropFirst())
        }

        guard !stripped.isEmpty else { return nil }

        let chars = Array(stripped)

        // Memoization: track (charIndex, parentDir) states known to fail.
        var failedStates = Set<String>()
        let deadline = Date().addingTimeInterval(Self.resolveTimeoutSeconds)

        return buildPath(
            chars: chars,
            charIndex: 0,
            parentDir: "/",
            componentSoFar: "",
            failedStates: &failedStates,
            deadline: deadline
        )
    }

    /// Recursively resolve the encoded string character-by-character.
    ///
    /// - Parameters:
    ///   - chars: The full encoded string as a character array (leading `-` already stripped).
    ///   - charIndex: Current position in the character array.
    ///   - parentDir: The validated directory path built so far.
    ///   - componentSoFar: The partial path component being assembled.
    ///   - failedStates: Memoization set of `"charIndex:parentDir"` keys that lead nowhere.
    ///   - deadline: Absolute time after which we stop searching.
    /// - Returns: A valid filesystem path, or `nil`.
    private func buildPath(
        chars: [Character],
        charIndex: Int,
        parentDir: String,
        componentSoFar: String,
        failedStates: inout Set<String>,
        deadline: Date
    ) -> String? {
        // Check timeout
        if Date() > deadline { return nil }

        // Base case: consumed all characters
        if charIndex >= chars.count {
            if componentSoFar.isEmpty {
                return fileManager.fileExists(atPath: parentDir) ? parentDir : nil
            }
            let finalPath = (parentDir as NSString).appendingPathComponent(componentSoFar)
            return fileManager.fileExists(atPath: finalPath) ? finalPath : nil
        }

        // Memoization key: captures position in the encoded string and which
        // validated directory we've reached. The componentSoFar is deliberately
        // excluded — we only memo when entering a new directory (componentSoFar
        // is empty after a `/` split), which is the key branching point.
        if componentSoFar.isEmpty {
            let stateKey = "\(charIndex):\(parentDir)"
            if failedStates.contains(stateKey) { return nil }
        }

        let ch = chars[charIndex]

        if ch != "-" {
            // Not a dash — this character is unambiguous, append it.
            let result = buildPath(
                chars: chars,
                charIndex: charIndex + 1,
                parentDir: parentDir,
                componentSoFar: componentSoFar + String(ch),
                failedStates: &failedStates,
                deadline: deadline
            )
            if result == nil && componentSoFar.isEmpty {
                failedStates.insert("\(charIndex):\(parentDir)")
            }
            return result
        }

        // The character is `-`. Try each possible interpretation.

        // Option 1: This dash is a `/` (path component boundary).
        // Only valid if we have accumulated a component name.
        if !componentSoFar.isEmpty {
            let candidateDir = (parentDir as NSString).appendingPathComponent(componentSoFar)
            var isDir: ObjCBool = false
            if fileManager.fileExists(atPath: candidateDir, isDirectory: &isDir),
               isDir.boolValue {
                if let result = buildPath(
                    chars: chars,
                    charIndex: charIndex + 1,
                    parentDir: candidateDir,
                    componentSoFar: "",
                    failedStates: &failedStates,
                    deadline: deadline
                ) {
                    return result
                }
            }
        }

        // Option 2: This dash represents `-`, ` `, or `.` within a component.
        for replacement in Self.replacementChars {
            if let result = buildPath(
                chars: chars,
                charIndex: charIndex + 1,
                parentDir: parentDir,
                componentSoFar: componentSoFar + String(replacement),
                failedStates: &failedStates,
                deadline: deadline
            ) {
                return result
            }
        }

        // Record failure for memoization (only at directory boundaries)
        if componentSoFar.isEmpty {
            failedStates.insert("\(charIndex):\(parentDir)")
        }

        return nil
    }
}

File: electron/src/main/services/ProjectDiscovery.ts
// ─── Project Discovery ──────────────────────────────────────────────────────
//
// Discovers Claude Code projects by scanning ~/.claude/projects/ and resolving
// the encoded directory names back to real filesystem paths.
//
// Encoded names replace `/`, `-`, ` `, and `.` with `-`, making them ambiguous.
// We use recursive backtracking with filesystem validation to resolve them.
//

import * as fs from 'fs'
import * as path from 'path'
import { homedir } from 'os'
import Database from 'better-sqlite3'
import { ProjectDAO } from '../database/dao/ProjectDAO'
import { parseSessionFile } from './SessionParser'
import { SessionDAO } from '../database/dao/SessionDAO'

export interface DiscoveredProject {
  encodedName: string
  resolvedPath: string | null
  name: string
  sessionFiles: string[]
}

/**
 * Resolve an encoded Claude project directory name back to its real filesystem path.
 *
 * The encoding replaces `/`, `-`, ` `, and `.` all with `-`.
 * This is ambiguous, so we try all possible interpretations using backtracking
 * and validate against the actual filesystem.
 *
 * @param encoded - The encoded directory name (e.g., `-Users-nicknorris-Documents-my-project`)
 * @param timeoutMs - Maximum time to spend on resolution (default 500ms)
 * @returns The resolved absolute path, or null if resolution fails
 */
export function resolvePath(encoded: string, timeoutMs = 500): string | null {
  const isWindows = process.platform === 'win32'

  let rootDir: string
  let chars: string

  if (isWindows) {
    // Windows: encoded names look like `C--Users-mcpme-...`
    // The drive letter + first dash represents `C:\`
    const winMatch = encoded.match(/^([A-Za-z])-(-.*)?$/)
    if (!winMatch) return null
    const driveLetter = winMatch[1].toUpperCase()
    rootDir = `${driveLetter}:\\`
    // The rest after `X-` — strip the leading `-` that represents `\`
    chars = winMatch[2] ? winMatch[2].slice(1) : ''
  } else {
    // Unix: encoded names start with `-` representing the leading `/`
    if (!encoded.startsWith('-')) return null
    rootDir = '/'
    chars = encoded.slice(1)
  }

  if (chars.length === 0) return null

  const deadline = Date.now() + timeoutMs
  const failed = new Set<string>()

  function backtrack(charIndex: number, parentDir: string, current: string): string | null {
    // Check timeout
    if (Date.now() > deadline) return null

    // If we've consumed all characters, check if the full path exists
    if (charIndex === chars.length) {
      const fullPath = path.join(parentDir, current)
      try {
        fs.statSync(fullPath)
        return fullPath
      } catch {
        return null
      }
    }

    const ch = chars[charIndex]

    if (ch !== '-') {
      // Non-dash character: append to current component
      return backtrack(charIndex + 1, parentDir, current + ch)
    }

    // Dash character: try multiple interpretations
    const memoKey = `${charIndex}:${parentDir}:${current}`
    if (failed.has(memoKey)) return null

    // 1. Try as path separator — current component must exist as a directory
    if (current.length > 0) {
      const dirPath = path.join(parentDir, current)
      try {
        const stat = fs.statSync(dirPath)
        if (stat.isDirectory()) {
          const result = backtrack(charIndex + 1, dirPath, '')
          if (result) return result
        }
      } catch {
        // Directory doesn't exist, skip this interpretation
      }
    }

    // 2. Try as literal `-`
    const r1 = backtrack(charIndex + 1, parentDir, current + '-')
    if (r1) return r1

    // 3. Try as literal ` `
    const r2 = backtrack(charIndex + 1, parentDir, current + ' ')
    if (r2) return r2

    // 4. Try as literal `.`
    const r3 = backtrack(charIndex + 1, parentDir, current + '.')
    if (r3) return r3

    failed.add(memoKey)
    return null
  }

  return backtrack(0, rootDir, '')
}

/**
 * Discover all Claude Code projects from ~/.claude/projects/.
 *
 * Lists encoded project directories, attempts to resolve each to a real
 * filesystem path, and collects session file information.
 */
export function discoverProjects(): DiscoveredProject[] {
  const claudeProjectsDir = path.join(homedir(), '.claude', 'projects')

  let entries: string[]
  try {
    entries = fs.readdirSync(claudeProjectsDir)
  } catch {
    return []
  }

  const projects: DiscoveredProject[] = []

  for (const entry of entries) {
    // Skip the bare `-` directory (represents root `/`)
    if (entry === '-') continue

    const entryPath = path.join(claudeProjectsDir, entry)

    // Only process directories
    try {
      if (!fs.statSync(entryPath).isDirectory()) continue
    } catch {
      continue
    }

    // Attempt to resolve the encoded name to a real path
    const resolvedPath = resolvePath(entry)

    // Derive a human-readable name
    const name = resolvedPath
      ? path.basename(resolvedPath)
      : entry.split('-').filter(Boolean).pop() || entry

    // Collect .jsonl session files
    let sessionFiles: string[] = []
    try {
      sessionFiles = fs
        .readdirSync(entryPath)
        .filter((f) => f.endsWith('.jsonl') && isUUID(f.replace('.jsonl', '')))
    } catch {
      // If we can't read the directory, just report 0 sessions
    }

    projects.push({
      encodedName: entry,
      resolvedPath,
      name,
      sessionFiles,
    })
  }

  return projects
}

/**
 * Sync discovered projects with the database.
 *
 * - Creates new projects for newly discovered Claude directories
 * - Updates claudeProject path for existing projects matched by filesystem path
 * - Does NOT delete projects that are no longer on disk
 */
export function syncProjectsWithDatabase(
  db: Database.Database,
  discovered: DiscoveredProject[]
): void {
  const projectDAO = new ProjectDAO(db)
  const existingProjects = projectDAO.list()

  for (const disc of discovered) {
    if (!disc.resolvedPath) continue

    // Check if a project with this filesystem path already exists
    const existing = existingProjects.find((p) => p.path === disc.resolvedPath)

    if (existing) {
      // Update claudeProject if not already set
      if (!existing.claudeProject || existing.claudeProject !== disc.encodedName) {
        projectDAO.update(existing.id, { claudeProject: disc.encodedName })
      }
    } else {
      // Create new project
      projectDAO.create({
        name: disc.name,
        path: disc.resolvedPath,
        claudeProject: disc.encodedName,
      })
    }
  }
}

/**
 * Import all sessions for a specific project from its Claude directory.
 *
 * Reads each .jsonl file, parses it with SessionParser, and upserts
 * the session into the database.
 *
 * Skips re-parsing if session already exists AND has token data.
 */
export function importProjectSessions(
  db: Database.Database,
  projectId: string,
  encodedName: string
): number {
  const claudeDir = path.join(homedir(), '.claude', 'projects', encodedName)
  const sessionDAO = new SessionDAO(db)

  let files: string[]
  try {
    files = fs
      .readdirSync(claudeDir)
      .filter((f) => f.endsWith('.jsonl') && isUUID(f.replace('.jsonl', '')))
  } catch {
    return 0
  }

  let imported = 0

  for (const file of files) {
    const sessionId = file.replace('.jsonl', '')

    // Skip if session already exists with token data
    const existing = sessionDAO.getById(sessionId)
    if (existing && (existing.inputTokens > 0 || existing.outputTokens > 0)) {
      continue
    }

    let content: string
    try {
      content = fs.readFileSync(path.join(claudeDir, file), 'utf-8')
    } catch {
      continue
    }

    const parsed = parseSessionFile(content, sessionId)

    if (existing) {
      // Update existing session
      sessionDAO.update(sessionId, {
        endedAt: parsed.endedAt ?? undefined,
        messageCount: parsed.messageCount,
        toolUseCount: parsed.toolUseCount,
        filesChanged: parsed.filesChanged.length > 0 ? JSON.stringify(parsed.filesChanged) : undefined,
        inputTokens: parsed.inputTokens,
        outputTokens: parsed.outputTokens,
        cacheCreationTokens: parsed.cacheCreationTokens,
        cacheReadTokens: parsed.cacheReadTokens,
      })
    } else {
      // Create new session
      sessionDAO.create({
        id: sessionId,
        projectId,
        slug: parsed.slug ?? undefined,
        startedAt: parsed.startedAt ?? undefined,
        model: parsed.model ?? undefined,
        gitBranch: parsed.gitBranch ?? undefined,
      })
      // Then update with the rest of the parsed data
      sessionDAO.update(sessionId, {
        endedAt: parsed.endedAt ?? undefined,
        messageCount: parsed.messageCount,
        toolUseCount: parsed.toolUseCount,
        filesChanged: parsed.filesChanged.length > 0 ? JSON.stringify(parsed.filesChanged) : undefined,
        inputTokens: parsed.inputTokens,
        outputTokens: parsed.outputTokens,
        cacheCreationTokens: parsed.cacheCreationTokens,
        cacheReadTokens: parsed.cacheReadTokens,
      })
    }

    imported++
  }

  return imported
}

/**
 * Check if a string is a valid UUID v4 format.
 */
function isUUID(str: string): boolean {
  return /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i.test(str)
}
